src/scores/continuous/nse_impl.py

```diff
@@ -0,0 +1,494 @@
+    #: whether the input data is angular, used implicitly by `mse`
+    is_angular: bool = False
+
+    def __post_init__(self):
+        """
+        Run checks before creating final object
+            - runtime type assertions
+            - check weights
+            - dimension checks
+        """
+        # safety: dev/testing: runtime assertions on arguments
+        assert (
+            isinstance(self.fcst, xr.Dataset)
+            and isinstance(self.obs, xr.Dataset)
+            # These are optionals (i.e. skipped if None):
+            and (self.weights is None or isinstance(self.weights, xr.Dataset))
+            and (self.reduce_dims is None or is_flexibledimensiontypes(self.reduce_dims))
+            and (self.preserve_dims is None or is_flexibledimensiontypes(self.preserve_dims))
+            and isinstance(self.is_angular, bool)
+        )
+
+        # check weights if its provided
+        if self.weights is not None:
+            check_weights(self.weights)
+
+        # do any dimension checks and store the final reduced (gathered) dimensions
+        self.reduce_dims = NseUtils.check_and_gather_dimensions(
+            fcst=self.fcst,
+            obs=self.obs,
+            weights=self.weights,
+            reduce_dims=self.reduce_dims,
+            preserve_dims=self.preserve_dims,
+        )
+
+    # The following methods are defined as "properties", since a NseScore object should
+    # not be mutable so any computation should also only act as a "getter"
+
+    @property
+    def nse(self) -> xr.Dataset:
+        """
+        Actual ``nse`` computation happens here.
+
+        Divide by zero error is ignored as it will be already raised by call to
+        ``obs_variance`` computation.
+        """
+        with np.errstate(divide="ignore"):
+            ret_nse: xr.Dataset = 1.0 - (self.fcst_error / self.obs_variance)
+            return ret_nse
+
+    @property
+    def fcst_error(self) -> XarrayLike:
+        """
+        Calculates fcst error, essentially : ``scores.mse(fcst, obs)``
+        """
+        return self._mse_callback(self.fcst, self.obs)
+
+    @property
+    def obs_variance(self) -> XarrayLike:
+        """
+        Calculates obs variance, essentially: ``scores.mse(obs_mean, obs)``
+        """
+        # compute mean along reduce dims - this does not need to be weighted
+        obs_mean: XarrayLike = self.obs.mean(dim=self.reduce_dims)
+        # using `mse` since its compatible with weights and angular data
+        obs_var: XarrayLike = self._mse_callback(obs_mean, self.obs)
+        # check for 0 obs variance: divide by zero condition
+        NseUtils.check_nonzero_obsvar(obs_var)
+
+        return obs_var
+
+    @property
+    def _mse_callback(self):
+        return functools.partial(
+            scores.continuous.mse,
+            reduce_dims=self.reduce_dims,
+            weights=self.weights,
+            is_angular=self.is_angular,
+        )
```

Comment:
Might as well make this callback more useful and typehint/assert that it returns a dataset.

I could get rid of the property tag as well, and not return a callback, but I think keeping everything consistently as read-only properties, is not terrible - even if it doesn't seem "natural" to have a read-only callback.

Noting the alternative where class methods (or their instantiated objects) can be mutated, i.e. `obj_mse_score.mse_callback = None` - is a thing that I don't want allowed. I'm not sure that property is the _best way_ to do it, but it is _a way_ to do it

.. note::

This is not a drop-in suggestion, just psuedo-code

```suggestion
    @property
    def _mse_callback(self) -> Callable[[xr.Dataset, xr.Dataset], xr.Dataset]:
        def _mse_return_dataset(_ds_x: xr.Dataset, _ds_y: xr.Dataset) -> xr.Dataset:
            ret: xr.Dataset = scores.continous.mse(
                fcst=_ds_x,
                obs=ds_y,
                ...  # insert non-changing members here
            )
            # safety: dev/testing only - should only return dataset
            assert(ret, xr.Dataset)
        return _mse_return_dataset
```

Reply:
I'm having a re-think about properties - I think there's a better way to handle this.

Reply:
New way now uses named tuples as compound structures returned by the main scoring functions instead of relying on property types.